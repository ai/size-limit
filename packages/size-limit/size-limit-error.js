const MESSAGES = {
  noPackage: () =>
    'Size Limit didn’t find *package.json*. ' +
    'Create npm package and run Size Limit there.',
  unknownArg: arg =>
    `Unknown argument *${arg}*. Check command for typo and read docs.`,
  argWithoutWebpack: arg =>
    `Argument *--${arg}* works only with *@size-limit/webpack* plugin` +
    (arg === 'why' ? '. You can add Bundle Analyzer to you own bundler.' : ''),
  argWithoutAnotherArg: (arg, anotherArg) =>
    `Argument *--${arg}* works only with *--${anotherArg}* argument`,
  argWithoutParameter: (arg, parameter) =>
    `Missing parameter *${parameter}* for *--${arg}* argument`,
  noConfig: () => 'Create Size Limit config in *package.json*',
  noArrayConfig: () => 'Size Limit config must contain *an array*',
  emptyConfig: () => 'Size Limit config must *not be empty*',
  noObjectCheck: () => 'Size Limit config array should contain *only objects*',
  pathNotString: () =>
    'The *path* in Size Limit config ' +
    'must be *a string* or *an array of strings*',
  entryNotString: () =>
    'The *entry* in Size Limit config ' +
    'must be *a string* or *an array of strings*',
  pluginlessConfig: (opt, mod) =>
    `Config option *${opt}* needs *@size-limit/${mod}* plugin`,
  multiPluginlessConfig: (opt, mod1, mod2) =>
    `Config option *${opt}* needs *@size-limit/${mod1}* ` +
    `or *@size-limit/${mod2}* plugin`,
  timeWithoutPlugin: () => 'Add *@size-limit/time* plugin to use time limit',
  unknownOption: opt =>
    `Unknown option *${opt}* in config. Check Size Limit docs and version.`,
  missedPlugin: mod => `Add *@size-limit/${mod}* plugin to Size Limit`,
  unknownEntry: entry =>
    `Size Limit didn’t find *${entry}* entry in custom Webpack config`,
  brotliUnsupported: () =>
    'Update your Node.js to version >= v11.7.0 to use Brotli',
  cmdError: (cmd, error) => (error ? `${cmd} error: ${error}` : `${cmd} error`),
  bundleDirNotEmpty: dir =>
    `The directory *${dir}* is not empty. ` +
    'Pass *--clean-dir* if you want to remove it'
}

const ADD_CONFIG_EXAMPLE = {
  noConfig: true,
  emptyConfig: true,
  noObjectCheck: true,
  noArrayConfig: true,
  pathNotString: true
}

class SizeLimitError extends Error {
  constructor (type, ...args) {
    super(MESSAGES[type](...args))
    this.name = 'SizeLimitError'
    if (ADD_CONFIG_EXAMPLE[type]) {
      this.example =
        '  "size-limit": [\n' +
        '    {\n' +
        '      "path": "dist/bundle.js",\n' +
        '      "limit": "10 KB"\n' +
        '    }\n' +
        '  ]\n'
    }
  }
}

module.exports = SizeLimitError
