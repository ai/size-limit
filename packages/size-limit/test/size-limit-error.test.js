import { expect, it } from 'vitest'

import createReporter from '../create-reporter'
import { SizeLimitError } from '../size-limit-error'

function print(err) {
  let stderr = ''
  let process = {
    stderr: {
      write(str) {
        stderr += str
      }
    }
  }
  let processor = createReporter(process, false)
  processor.error(err)
  return stderr
}

it('has mark', () => {
  let err = new SizeLimitError('noPackage')
  expect(err.name).toBe('SizeLimitError')
})

it('has start', () => {
  let err = new SizeLimitError('noPackage')
  expect(err.stack).toContain('size-limit-error.test.js')
})

it('has message', () => {
  let err = new SizeLimitError('noPackage')
  expect(err.message).toContain('Create npm package')
})

it('has error for unknown option', () => {
  let err = new SizeLimitError('missedPlugin', 'webpack', 'file')
  expect(print(err)).toMatchSnapshot()
})

it('has error for CLI error', () => {
  let err = new SizeLimitError(
    'cmdError',
    'cli-tool',
    'Module not found\n  @ multi ./dual-publish-tmp/index.js index[0]'
  )
  expect(print(err)).toMatchSnapshot()
})

it('has error for CLI error without output', () => {
  let err = new SizeLimitError('cmdError', 'cli-tool')
  expect(print(err)).toMatchSnapshot()
})

it('has error for unknown entry', () => {
  let err = new SizeLimitError('unknownEntry', 'admin')
  expect(print(err)).toMatchSnapshot()
})
