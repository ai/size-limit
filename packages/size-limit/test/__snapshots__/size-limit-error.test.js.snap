// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`has error for CLI error 1`] = `
"[41m[30m ERROR [39m[49m [31mcli-tool error: Module not found
  @ multi ./dual-publish-tmp/index.js index[0][39m
"
`;

exports[`has error for CLI error without output 1`] = `
"[41m[30m ERROR [39m[49m [31mcli-tool error[39m
"
`;

exports[`has error for unknown entry 1`] = `
"[41m[30m ERROR [39m[49m [31mSize Limit didnâ€™t find [33madmin[31m entry in custom Webpack config[39m
"
`;

exports[`has error for unknown option 1`] = `
"[41m[30m ERROR [39m[49m [31mAdd [33m@size-limit/webpack[31m plugin to Size Limit[39m
"
`;
